sklearn/impute/_base.py

```diff
@@ -363,6 +363,26 @@ def _validate_input(self, X, in_fit):
                 "with an object dtype.".format(X.dtype)
             )
 
+        if sp.issparse(X) and self.missing_values == 0:
+            # missing_values = 0 not allowed with sparse data as it would
+            # force densification
+            raise ValueError(
+                "Imputation not possible when missing_values "
+                "== 0 and input is sparse. Provide a dense "
+                "array instead."
+            )
+
+        if self.strategy == "constant" and self.fill_value is not None:
+            if in_fit:
+                fill_value = self.fill_value
+            else:
+                fill_value = self.statistics_[0]
+            if not np.can_cast(type(fill_value), X.dtype, "same_kind"):
+                raise ValueError(
+                    f"fill_value={fill_value} cannot be cast to the input dtype"
+                    f" {X.dtype}"
+                )
```

Comment:
Let's specialize the error message depending if we are in `fit` or `transform`.
Also `self.fill_value is not None` should only be check if in `fit`. Otherwise we use the array of statistics created at `fit`. This could be potentially wrong at transform if the input dtype change for some reason.

```suggestion
        if self.strategy == "constant":
            # ensure that we can safely cast the value to fill the missing values with
            # if needed
            if in_fit and self.fill_value is not None:
                fill_value_dtype = type(self.fill_value)
                err_msg = (
                    f"fill_value={self.fill_value} (of type {fill_value_dtype}) cannot "
                    f"be cast to the input data that is {X.dtype}. Make sure that both "
                    "dtypes are of the same kind."
                )
            else:
                fill_value_dtype = self.statistics_.dtype
                err_msg = (
                    f"The dtype of the filling statistic (i.e. {fill_value_dtype}) "
                    f"cannot be cast to the input data that is {X.dtype}. Make sure "
                    "that the dtypes of the input data is of the same kind between "
                    "fit and transform."
                )
            if not np.can_cast(fill_value_dtype, X.dtype, "same_kind"):
                raise ValueError(err_msg)
```